File: .tmp_py4013/batch_18_dump/3984_divisible_game.js
// LeetCode 3984 - Divisible Game
// https://leetcode.com/problems/divisible-game/
var divisibleGame = function(nums) {
        let candidates = new Set();
        candidates.push(2);
        for (const value of nums) {
            for (let divisor = 2; divisor * divisor <= value; divisor++) {
                if (value % divisor != 0) continue;
                candidates.push(divisor);
                candidates.push(value / divisor);
            }
            if (value > 1) candidates.push(value);
        }
        let bestScore = -(1 << 62);
        let bestK = 0;
        for (const k of candidates) {
            let ending = 0, score = 0;
            for (let i = 0; i < nums.length; i++) {
                let value = nums[i];
                let contribution = -(value);
                if (value % k == 0) contribution = value;
                if (i == 0 || ending + contribution < contribution) ending = contribution;
                else ending += contribution;
                if (i == 0 || ending > score) score = ending;
            }
            if (score > bestScore || (score == bestScore && k < bestK)) {
                bestScore = score;
                bestK = k;
            }
        }
        let mod = 1000000007;
        let answer = (bestScore % mod) * bestK % mod;
        if (answer < 0) answer += mod;
        return answer;
    
};

File: .tmp_rs4013/rust_check22/3984_divisible_game.rs
struct Solution;
// LeetCode 3984 - Divisible Game
// https://leetcode.com/problems/divisible-game/

use std::collections::HashSet;

impl Solution {
    pub fn divisible_game(nums: Vec<i32>) -> i32 {
        let mut candidates: HashSet<i32> = HashSet::new();
        candidates.insert(2);
        for &value in &nums {
            let mut divisor = 2;
            while divisor * divisor <= value {
                if value % divisor == 0 {
                    candidates.insert(divisor);
                    candidates.insert(value / divisor);
                }
                divisor += 1;
            }
            if value > 1 {
                candidates.insert(value);
            }
        }
        let mut best_score = -(1i64 << 62);
        let mut best_k = 0;
        for &k in &candidates {
            let mut ending = 0i64;
            let mut score = 0i64;
            for (i, &value) in nums.iter().enumerate() {
                let mut contribution = -(value as i64);
                if value % k == 0 {
                    contribution = value as i64;
                }
                if i == 0 || ending + contribution < contribution {
                    ending = contribution;
                } else {
                    ending += contribution;
                }
                if i == 0 || ending > score {
                    score = ending;
                }
            }
            if score > best_score || (score == best_score && k < best_k) {
                best_score = score;
                best_k = k;
            }
        }
        const MOD: i64 = 1_000_000_007;
        let mut answer = (best_score % MOD) * best_k as i64 % MOD;
        if answer < 0 {
            answer += MOD;
        }
        answer as i32
    }
}

fn main() {}
